core/code/ornaments.js

```diff
@@ -13,52 +13,120 @@
-    this._beacons = layerGroup();
-    this._frackers = layerGroup();
-    window.layerChooser.addOverlay(this._layer, 'Ornaments');
-    window.layerChooser.addOverlay(this._beacons, 'Beacons');
-    window.layerChooser.addOverlay(this._frackers, 'Frackers');
+    this.load();
+
+    this.layers = {};
+    this.layers._layer = window.ornaments.layerGroup();
+    this.layers._excluded = window.ornaments.layerGroup(); // to keep excluded ornaments in an own layer
+
+    window.layerChooser.addOverlay(this.layers._layer, 'Ornaments');
+    window.layerChooser.addOverlay(this.layers._excluded, 'Excluded ornaments', {default: false});
+
+    $('<a>')
+      .html('Ornaments Opt')
+      .attr({
+        id: 'ornaments-toolbox-link',
+        title: 'Edit ornament exclusions',
+        accesskey: 'o'
+      })
+      .click(window.ornaments.ornamentsOpt)
+      .appendTo('#toolbox');
+  },
+  createLayer: function (layerID) {
+    window.ornaments.layers[layerID] = window.ornaments.layerGroup();
+    window.layerChooser.addOverlay(window.ornaments.layers[layerID], layerID);
   },
 
   addPortal: function (portal) {
     this.removePortal(portal);
-
     var ornaments = portal.options.data.ornaments;
     if (ornaments && ornaments.length) {
       this._portals[portal.options.guid] = ornaments.map(function (ornament) {
-        var layer = this._layer;
-        if (ornament.startsWith('pe')) {
-          layer = ornament === 'peFRACK'
-            ? this._frackers
-            : this._beacons;
-        }
+        var layer = this.layers._layer;
+        var opacity = this.OVERLAY_OPACITY;
         var size = this.OVERLAY_SIZE;
+        var anchor = [size / 2, size / 2];
+        var iconUrl = '//commondatastorage.googleapis.com/ingress.com/img/map_icons/marker_images/' + ornament + '.png';
+
+        if (!this.knownOrnaments[ornament]) {
+          this.knownOrnaments[ornament]=false;
+        }
+
+        if (typeof (window.ornaments.icon[ornament]) !== 'undefined') {
+          if (window.ornaments.icon[ornament].layer) {
+            if (window.ornaments.layers[window.ornaments.icon[ornament].layer] === undefined){
+              console.log ('Add missing layer: ',window.ornaments.icon[ornament].layer);
+              window.ornaments.createLayer(window.ornaments.icon[ornament].layer);
+            }
+            layer =  window.ornaments.layers[window.ornaments.icon[ornament].layer];
+          }
+          opacity = 1;
+          if (window.ornaments.icon[ornament].url) {
+            iconUrl = window.ornaments.icon[ornament].url;
+            if (window.ornaments.icon[ornament].offset) {
+              switch (window.ornaments.icon[ornament].offset) {
+              case 1:
+                anchor = [size / 2, size];
+                break;
+              case 0:
+                anchor = [size / 2, size / 2];
+                break;
+              case -1:
+                anchor = [size / 2, - size ];
+              }
+            }
```

Comment:
this `.offset` in `-1,0,1` looks oddly specific and restrictive to me.
why not having `.offset` an array instead default value  `[size/2,size/2]` ?
if a plugin wants an icon with some offset, then it defines `.offset: [dx, dy]`

Reply:
dx and dy would no longer calculate from the image's size. That means, whenever we change the (size of) the icon we would also need to adjust these parameters. We have just moved the images to iitc.app and that would cause a lot of testing. IMHO if you really need to dislocate an icon from the offered offsets you are free to simply draw a larger box and dislocate in that box.
![down](https://user-images.githubusercontent.com/25123137/178440653-5319b681-3fe2-496f-afd4-1d78ed92365d.png)

Reply:
> dx and dy would no longer calculate from the image's size.

the size of the pictures is a constant set as `window.ornaments.OVERLAY_SIZE`, that is available from any plugin.

> whenever we change the (size of) the icon we would also need to adjust these parameters.

you would need to change the value of `.OVERLAY_SIZE` and the icons of the plugin, so this offset doesn't really help
Also, all ornaments have the same fixed size due to how stock intel works, that's something out of our control.

> you are free to simply draw a larger box and dislocate in that box.

the code doesn't support different ornament size (see above)

Reply:
case -1 = ,offset=[0.5,-1] =>  ` anchor = [size * offset[0], size * offset[1]];`

But this is excatly what I meant when I said: this is code that should go into the plugin and not in the core.
The plugin knows excatly how to display stuff. The ornaments class is just doing assumptation what could be required based on his own plugin.

as said in another chat: what about mouse events? animations? tooltips? (opacity?)

Reply:
> as said in another chat: what about mouse events? animations? tooltips? (opacity?)

current ornament layer is not suited for mouse event (so tooltips), because of how leaflet and canvas works.
this events can be added to the portal itself instead of the ornaments.

But, we can discuss if using a specific canvas for each ornaments layer is more clever (performance wise) than drawing the ornaments directly on the canvas layer.

Reply:
> the size of the pictures is a constant set as `window.ornaments.OVERLAY_SIZE`, that is available from any plugin.

yep, missed that. 

But still: I think it is allowed to define certain limits to what kind of images we accept and how we display them (KISS). Initially I only wanted to be able to "hide" ornaments and now the whole discussion is now about what fancy things can be done with the images/icons/ornaments shown and that we should not limit that at all, where actually I have not yet seen a request nor a reasonable use-case.

IITC is limited in so may things by the poor data we receive from stock intel, trying to improve things by adding fancy animations or similar to "no data" is imho only serving DEVs not the user. 

my2ct

Reply:
you are the one introducing a new offset feature, which is likely to be changed because too restrictive as it is imo, and a future change implies backward compatibility or breaking plugins... more work afterwards...
 
@McBen suggestion looks kiss enough

```suggestion
            if (window.ornaments.icon[ornament].offset) {
              var offset = window.ornaments.icon[ornament].offset;
              anchor = [size * offset[0], size * offset[1]];
            }
```

Reply:
going to accept that, but should we add some limits to the value just for safety?  -1..1 will keep it close to the portal.

Reply:
I don't think validation is necessary. It is enough to describe the limits of possible values. If the developer wants to specify values outside the boundaries, that's up to him

Reply:
Accepted to new PR